### PR TITLE
Feature/rework

### PR DESCRIPTION
**Pull request by naxhh**
The first major re-write was a big step in the right direction.
But I still feel there was more job to be done.

This re-write pretends to simplify even more the roles and abstract them of the logic behind when a role is added and where not.

So the first step was to remove BaseRole because was performing RoleManager logic.
I've moved that logic to the manager itself. So now is the manager who decides if a role should be added to the bundle or not.

The roles are only able to specify they requirements.
The RoleInterface was renamed to just `Role`
And two new interfaces where created.

`RoleWithDependencies ` provides a method to specify which roles this role depends on. This is inside a new interface so this method is optional to be implemented by any `Role`
Examples: Composer and Xdebug roles that depend on PHP one

`RoleValuesTransformer` provides a way to transform or add new values to the role configuration.
This is required right now by the PHP role and the VagrantLocal role.
This two roles had a special requirement of having the VagrantBundle. So I've added that to the signature.

This is the part I don't really like. Because from my point of view PHP role should be able to do his "thing" without this. And VagrantLocal should not be a role at all, but a "target"

But this will come in later rewrites when I have a better idea of how to do this :)

Regards and thanks for your time looking at this.

ping @phansible/collaborators

### CODE REVIEW

File: src/Phansible/Role.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Phansible;
+
+interface Role
```
Comment (sndpl) [inline src/Phansible/Role.php]: I would follow the naming conventions from the php-fig group: https://github.com/php-fig/fig-standards/blob/master/bylaws/002-psr-naming-conventions.md  and rename it to RoleInterface. Same for the new interface you created.

File: src/Phansible/Roles/Nginx.php [hunk 2]
```diff
@@ -2,11 +2,29 @@
...
+    {
+        return 'nginx';
+    }
+
+    public function getRole()
+    {
+        return 'nginx';
+    }
+
+    public function requiresRoles()
```
Comment (sndpl) [inline src/Phansible/Roles/Nginx.php]: Why is this function here?

File: src/Phansible/RoleManager.php [hunk 3]
```diff
@@ -52,4 +41,49 @@ public function getInitialValues()
...
+            if ($role instanceof RoleValuesTransformer) {
+                $values = $role->transformValues($values, $vagrantBundle);
+            }
+
+            $vagrantBundle->getPlaybook()->addRole($role->getRole());
+            $vagrantBundle->getVarsFile()->addMultipleVars([$role->getSlug() => $values]);
+        });
+    }
+
+    private function willBeInstalled($roleSlug, $requestVars)
```
Comment (sndpl) [inline src/Phansible/RoleManager.php]: Missing docblock

File: src/Phansible/RoleManager.php [hunk 4]
```diff
@@ -52,4 +41,49 @@ public function getInitialValues()
         }
         return $initials;
     }
+
+    /**
+     * {@inheritdoc}
```
Comment (sndpl) [inline src/Phansible/RoleManager.php]: There is no base class to inherit docs

File: src/Phansible/RoleManager.php (hunk 3 above)
Comment (sndpl) [inline src/Phansible/RoleManager.php]: Missing docblock

File: src/Phansible/Role.php (hunk 1 above)
Comment (sndpl) [inline src/Phansible/Role.php]: Ignore the above comment: http://verraes.net/2013/09/sensible-interfaces/ :)

File: src/Phansible/Roles/Nginx.php (hunk 2 above)
Comment (naxhh) [inline src/Phansible/Roles/Nginx.php]: I just forget to remove it. I will re-check all of them :)

File: src/Phansible/RoleManager.php (hunk 4 above)
Comment (naxhh) [inline src/Phansible/RoleManager.php]: ups!

File: src/Phansible/RoleManager.php (hunk 3 above)
Comment (naxhh) [inline src/Phansible/RoleManager.php]: I'm not a fan of commenting everything. But If you think an explanation of the method is required i'll add it.

File: src/Phansible/Role.php (hunk 1 above)
Comment (naxhh) [inline src/Phansible/Role.php]: Good! I'm not a fan of adding "Interface"

**Comment by InFog** [comment]
IMO this is the right way to go. Code is really looking better and easier to understand.

This thing with Vagrant being not a role, but a target, is very nice! If we go this way we will be able to have different targets and not only local Vagrant development environments :+1:

**Comment by naxhh** [comment]
That's my main point of setting Vagrant as target :) nice to know you like it.
I would try to close this this weekend and PR changes to the open PR's so we can move on.

**Pull request merged**